R/civis_future.R

```diff
@@ -0,0 +1,167 @@
+                           required_resources = required_resources,
+                           docker_image_name = docker_image_name,
+                           docker_image_tag = docker_image_tag, ...)
+  structure(future, class = c("CivisFuture", class(future)))
+}
+
+#' @importFrom future run value resolved
+NULL
+
+#' @export
+run.CivisFuture <- function(future, ...) {
+  cargo <- c(expr = future$expr, envir = future$envir, packages = list(future$packages))
+  task_file_id <- write_civis_file(cargo)
+  runner_file_id <- upload_runner_script()
+  cmd <- make_docker_cmd(task_file_id, runner_file_id)
+  future$job <- scripts_post_containers(future$required_resources,
+                                        docker_command = cmd,
+                                        docker_image_name = future$docker_image_name,
+                                        docker_image_tag = future$docker_image_tag, ...)
+  future$job <- scripts_post_containers_runs(future$job$id)
+  future$state <- c("running")
+  future
+}
+
+#' @export
+value.CivisFuture <- function(future, ...) {
+  if (future$state == "created") {
+    future <- run(future)
+  }
+  if (!future$state %in% c("succeeded", "failed", "cancelled")) {
+    # get the run from the future object
+    tryCatch({
+      runs <- await(scripts_get_containers_runs, id = future$job$containerId, run_id = future$job$id)
+      future$state <- runs$state
+      future$run <- runs
+      future$value <- fetch_output(runs)
+    }, error = function(e) {
+      future$state <- "failed"
+      stop(e)
+    })
+  }
+  future$value
+}
+
+#' Cancel the evaluation of a CivisFuture.
+#' @param future CivisFuture object to be cancelled.
+#' @param ... unused for CivisFuture.
+#' @export
+cancel <- function(future, ...) {
+  UseMethod("cancel")
+}
+
+#' @export
+cancel.CivisFuture <- function(future, ...) {
+  scripts_delete_containers_runs(id = future$job$containerId, run_id = future$job$id)
+  future$state <- "cancelled"
+}
+
+#' @export
+resolved.CivisFuture <- function(future, ...){
+  future$state %in% c("succeeded", "failed", "cancelled")
+}
+
+#' @export
+fetch_logs.CivisFuture <- function(object, limit, ...){
+  logs <- scripts_list_containers_runs_logs(object$job$containerId, run_id = object$job$id, limit = limit)
+  format_scripts_logs(logs)
+}
+
+fetch_output <- function(run) {
+  outputs <- scripts_list_containers_runs_outputs(run$containerId, run$id)
+  # We only save one object in the run script.
+  output_file_id <- outputs[[1]][["objectId"]]
+  read_civis(output_file_id)
+}
+
+make_docker_cmd <- function(task_file_id, run_script_file_id) {
+  cmd <- "Rscript -e \"civis::download_civis(${run_script_file_id}, 'run.R')\" && Rscript run.R ${task_file_id}"
+  stringr::str_interp(cmd)
+}
+
+upload_runner_script <- function() {
+  path <- system.file("scripts", "r_remote_eval.R", package = "civis")
+  write_civis_file(path, name = "r_remote_eval.R")
+}
+
+# This is to get around R CMD check
```

Comment:
https://github.com/HenrikBengtsson/future/commit/da1d57c1a7fff940b633e5e93b20a959301c9c2b

Maybe we should just rely on the most recent version? Looks like `getGlobalsAndPackages` is now exported.